File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        // using pointers
        Arrays.sort(nums);

        List<List<Integer>> res = new ArrayList<>();

        for(int i = 0 ; i < nums.length ; i++)
        {
            if(nums[i] > 0)
            {
                break;
            }
            if(i > 0 && nums[i] == nums[i-1])
            {
                continue;
            }
            int l = i + 1;
            int r = nums.length - 1;
            int cursum = 0;

            while(l < r)
            {
                cursum = nums[i] + nums[l] + nums[r];

                if(cursum > 0)
                {
                    r -= 1;
                }
                else if(cursum < 0)
                {
                    l += 1;
                }
                else
                {
                    res.add(Arrays.asList(nums[i] , nums[l] , nums[r]));
                    l += 1;
                    r -= 1;

                    while(nums[l] == nums[l-1] && l < r)
                    {
                        l += 1;
                    }
                }
            }
        }
        return res;

        
    }
}

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        threesum = 0

        for i , a in enumerate(nums):
            if i > 0 and a == nums[i-1]:
                continue 

            left = i + 1
            right = len(nums) - 1

            while left < right:
                threesum = a + nums[left] + nums[right]

                if threesum < 0:
                    left += 1
                elif threesum > 0 :
                    right -= 1
                else:
                    res.append([a , nums[left] , nums[right]])

                    # SKIPPING DUPLICATES 
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                        
                    left += 1
                    right -= 1
                    
        return res

        